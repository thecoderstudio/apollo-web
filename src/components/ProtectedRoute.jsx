import React from 'react';
import { Route } from 'react-router-dom';
import { connect } from 'react-redux';

function ProtectedRoute ({
  component: Component,
  fallbackComponent: FallbackComponent,
  ...rest
}) {
  return (
    <Route
      {...rest}
      render={ (props) =>
          rest.authenticated ? <Component {...props} /> : <FallbackComponent {...props} />
      }
    />
  )
}


export default connect(
  ({ auth }) => ({ authenticated: auth.authenticated })
)(ProtectedRoute);
