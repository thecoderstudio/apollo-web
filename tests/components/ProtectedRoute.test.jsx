import React from 'react';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import ProtectedRoute from '../../src/components/ProtectedRoute';

const mockStore = configureStore([]);

function getComponent(store, role=undefined) {
  return renderer.create(
    <Provider store={store}>
      <MemoryRouter
        initialEntries={[{ pathname: '/', key: 'testKey' }]}
      >
        <ProtectedRoute exact path='/' component="auth" fallbackComponent="unauth" role={role} />
      </MemoryRouter>
    </Provider>
  );
}


describe('ProtectedRoute', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      authenticated: false,
      currentUser: {
        hasChangedInitialPassword: true
      }
    });
  });

  it('renders correctly', () => {
    const tree = getComponent(store).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('correctly handles authenticated', () => {
    let tree = getComponent(store).toJSON();
    expect(tree).toMatchSnapshot();

    store = mockStore({
      authenticated: true,
      currentUser: {
        hasChangedInitialPassword: true
      }
    });

    tree = getComponent(store).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('correctly handles authenticated and not password changed', () => {
    let tree = getComponent(store).toJSON();
    expect(tree).toMatchSnapshot();

    store = mockStore({
      authenticated: true,
      currentUser: {
        hasChangedInitialPassword: false
      }
    });

    tree = getComponent(store).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('correctly handlers matching roles', () => {
    store = mockStore({
      authenticated: true,
      currentUser: {
        username: 'admin',
        role: {
          name: 'admin'
        },
        hasChangedInitialPassword: true
      }
    });

    const tree = getComponent(store, 'admin').toJSON();
    expect(tree).toMatchSnapshot();

  });

  it('disallows non-matching roles', () => {
     store = mockStore({
      authenticated: true,
      currentUser: {
        username: 'admin',
        role: null
      },
      hasChangedInitialPassword: true
    });

    const tree = getComponent(store, 'admin').toJSON();
    expect(tree).toMatchSnapshot();
  });
});
