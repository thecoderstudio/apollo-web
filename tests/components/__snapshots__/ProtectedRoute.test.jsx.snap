// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProtectedRoute correctly handlers matching roles 1`] = `
<auth
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "testKey",
        "pathname": "/",
        "search": "",
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "hash": "",
      "key": "testKey",
      "pathname": "/",
      "search": "",
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {},
      "path": "/",
      "url": "/",
    }
  }
/>
`;

exports[`ProtectedRoute correctly handles authenticated 1`] = `
<unauth
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "testKey",
        "pathname": "/",
        "search": "",
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "hash": "",
      "key": "testKey",
      "pathname": "/",
      "search": "",
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {},
      "path": "/",
      "url": "/",
    }
  }
/>
`;

exports[`ProtectedRoute correctly handles authenticated 2`] = `
<auth
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "testKey",
        "pathname": "/",
        "search": "",
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "hash": "",
      "key": "testKey",
      "pathname": "/",
      "search": "",
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {},
      "path": "/",
      "url": "/",
    }
  }
/>
`;

exports[`ProtectedRoute disallows non-matching roles 1`] = `
<unauth
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "testKey",
        "pathname": "/",
        "search": "",
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "hash": "",
      "key": "testKey",
      "pathname": "/",
      "search": "",
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {},
      "path": "/",
      "url": "/",
    }
  }
/>
`;

exports[`ProtectedRoute renders correctly 1`] = `
<unauth
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "testKey",
        "pathname": "/",
        "search": "",
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "hash": "",
      "key": "testKey",
      "pathname": "/",
      "search": "",
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {},
      "path": "/",
      "url": "/",
    }
  }
/>
`;
